_posts/2024-08-06-how-to-talk-about-CHERI.markdown

```diff
@@ -0,0 +1,228 @@
+This is implemented with two properties on CHERIoT.
+The first is the two-bit information-flow-control model that CHERI has had since the very early days.
+This defines some capabilities as 'global' (and, conversely, the rest as local) and adds a store-local permission.
+You may store local capabilities only via a capability that has store-local permission.
+Global is not a permission (it doesn't authorise you to do anything) but behaves like one: it can be cleared, but not set.
+
+On a CHERIoT platform, the only memory that you will ever see with the store-local permission is the stack.
+This means that anything that you remove the global bit from can be held in registers or stored on the stack, but not stored on the heap or in globals.
+CHERIoT also adds a permit-load-global permission, so you can make this a deep property: no pointer loaded (at any depth of indirection) from this pointer may ever be stored on the globals or heap.
+This is combined with stack clearing in both directions of cross-compartment calls (with a bit of hardware assistance) to implement shallow or deep no-capture guarantees.
+A Rust compiler, for example, could use this to pass borrowed references to a C function and have a strong guarantee that the C code could not access the object beyond the function's return.
+
+## So you mean compilers and ISAs can be friends?
+
+None of these language-level properties come exclusively from the hardware.
+The hardware provides tools that the compiler can choose to use.
+The compiler can also choose not to use them when it doesn't need to.
+For safe languages, a lot of properties are guaranteed by construction and so don't need enforcing *within that language's code* (or within the safe subset, for a language like Rust), though a compiler for safety-critical systems may choose to use them for defence in depth against compiler bugs.
+In C, for example, we don't enforce stack temporal safety within a compartment because it's easy for static analysers to track this kind of bug when they can see all of the code and it's a better security-performance tradeoff to recommend that people aim the gun slightly away from their foot.
+We do enforce it at compartment boundaries, because we're crossing a security boundary.
+If you choose not to run a static analyser that checks for stack temporal safety issues, you can introduce bugs into your own code, but not other compartments.
+
+The same applies for CFI properties.
+CHERIoT provides a trusted stack for cross-compartment calls and a switcher that enforces a lot of properties on both the call and return path.
+Within a compartment, we provide forward and backwards return sentries for coarse-grained CFI (you can't confuse function pointers and return addresses).
+This provides a few guarantees at library boundaries.
+You cannot jump into the middle of a library function (easily, you may be able to if it spills its program counter to the stack and doesn't zero it on return).
```

Comment:
Rather "You can write library code that prevents a caller from jumping into the middle of a library function"